src/backend/distributed/executor/directed_acylic_graph_execution_logic.c

```diff
@@ -0,0 +1,228 @@
+{
+	HTAB *completedTasks = CreateTaskHashTable();
+
+	/* We only execute depended jobs' tasks, therefore to not execute */
+	/* top level tasks, we add them to the completedTasks. */
+	AddCompletedTasks(excludedTasks, completedTasks);
+	while (true)
+	{
+		List *curTasks = FindExecutableTasks(allTasks, completedTasks);
+		if (list_length(curTasks) == 0)
+		{
+			break;
+		}
+		ExecuteTaskListOutsideTransaction(ROW_MODIFY_NONE, curTasks,
+										  MaxAdaptiveExecutorPoolSize);
+
+		AddCompletedTasks(curTasks, completedTasks);
+		curTasks = NIL;
+	}
+}
+
+
+/*
+ * FindExecutableTasks finds the tasks that can be executed currently,
+ * which means that all of their dependencies are executed. If a task
+ * is already executed, it is not added to the result.
+ */
+static List *
+FindExecutableTasks(List *allTasks, HTAB *completedTasks)
+{
+	List *curTasks = NIL;
+	ListCell *taskCell = NULL;
+	TaskHashKey taskKey;
+
+	foreach(taskCell, allTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		FillTaskKey(&taskKey, task);
+
+		if (IsAllDependencyCompleted(task, completedTasks) &&
+			!IsTaskAlreadyCompleted(task, completedTasks))
+		{
+			curTasks = lappend(curTasks, task);
+		}
+	}
+
+	return curTasks;
+}
+
+
+/*
+ * AddCompletedTasks adds the givens tasks to completedTasks HTAB.
+ */
+static void
+AddCompletedTasks(List *curCompletedTasks, HTAB *completedTasks)
+{
+	ListCell *taskCell = NULL;
+	TaskHashKey taskKey;
+	bool found;
+
+	foreach(taskCell, curCompletedTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		FillTaskKey(&taskKey, task);
+		hash_search(completedTasks, &taskKey, HASH_ENTER, &found);
+	}
+}
+
+
+/*
+ * CreateTaskHashTable creates a HTAB with the necessary initialization.
+ */
+static HTAB *
+CreateTaskHashTable()
+{
+	uint32 hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT | HASH_COMPARE);
+	HASHCTL info = InitHashTableInfo();
+	return hash_create("citus task completed list (jobId, taskId)",
+					   64, &info, hashFlags);
+}
+
+
+/*
+ * IsTaskAlreadyCompleted returns true if the given task
+ * is found in the completedTasks HTAB.
+ */
+static bool
+IsTaskAlreadyCompleted(Task *task, HTAB *completedTasks)
+{
+	TaskHashKey taskKey;
+	bool found;
+
+	FillTaskKey(&taskKey, task);
+	hash_search(completedTasks, &taskKey, HASH_ENTER, &found);
+	return found;
+}
+
+
+/*
+ * IsAllDependencyCompleted return true if the given task's
+ * dependencies are completed.
+ */
+static bool
+IsAllDependencyCompleted(Task *targetTask, HTAB *completedTasks)
+{
+	ListCell *taskCell = NULL;
+	bool found = false;
+	TaskHashKey taskKey;
+
+	foreach(taskCell, targetTask->dependedTaskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		FillTaskKey(&taskKey, task);
+
+		hash_search(completedTasks, &taskKey, HASH_FIND, &found);
+		if (!found)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+static void
+FillTaskKey(TaskHashKey *taskKey, Task *task)
+{
+	taskKey->jobId = task->jobId;
+	taskKey->taskId = task->taskId;
```

Comment:
Why not just inline this instead of having the extra code?

TaskHashKey taskKey = { task->jobId, task->taskId };